File: eco-core/core-plugin/src/main/java/com/willfp/talismans/TalismansPlugin.java
package com.willfp.talismans;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.eco.core.items.CustomItem;
import com.willfp.talismans.command.CommandTalismans;
import com.willfp.talismans.display.TalismanDisplay;
import com.willfp.talismans.talismans.Talismans;
import com.willfp.talismans.talismans.util.BlockPlaceListener;
import com.willfp.talismans.talismans.util.DiscoverRecipeListener;
import com.willfp.talismans.talismans.util.TalismanChecks;
import com.willfp.talismans.talismans.util.TalismanCraftListener;
import com.willfp.talismans.talismans.util.WatcherTriggers;
import com.willfp.talismans.talismans.util.equipevent.SyncTalismanEquipEventTask;
import com.willfp.talismans.talismans.util.equipevent.TalismanEquipEventListeners;
import lombok.Getter;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

@SuppressWarnings("unused")
public class TalismansPlugin extends EcoPlugin {
    /**
     * Instance of the plugin.
     */
    @Getter
    private static TalismansPlugin instance;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public TalismansPlugin() {
        super(611, 9865, "&6", true);
        instance = this;
    }

    @Override
    protected void handleEnable() {
        this.getLogger().info(Talismans.values().size() + " Talismans Loaded");
        new CustomItem(this.getNamespacedKeyFactory().create("any_talisman"), test -> TalismanChecks.getTalismanOnItem(test) != null, Talismans.SHARPNESS_TALISMAN.getLevel(1).getItemStack()).register();
    }

    @Override
    protected void handleReload() {
        Talismans.values().forEach(talisman -> {
            HandlerList.unregisterAll(talisman);

            this.getScheduler().runLater(() -> {
                if (talisman.isEnabled()) {
                    this.getEventManager().registerListener(talisman);
                }
            }, 1);
        });
        SyncTalismanEquipEventTask.scheduleAutocheck(this);
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandTalismans(this)
        );
    }

    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new WatcherTriggers(this),
                new BlockPlaceListener(),
                new TalismanCraftListener(this),
                new TalismanEquipEventListeners(this),
                new DiscoverRecipeListener(this)
        );
    }

    @Override
    @Nullable
    protected DisplayModule createDisplayModule() {
        return new TalismanDisplay(this);
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/talismans/TalismansPlugin.kt
package com.willfp.talismans

import com.willfp.eco.core.bstats.EcoMetricsChart
import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.display.DisplayModule
import com.willfp.eco.core.items.Items
import com.willfp.libreforge.conditions.Conditions
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.registerSpecificHolderProvider
import com.willfp.libreforge.registerSpecificRefreshFunction
import com.willfp.libreforge.slot.SlotTypes
import com.willfp.talismans.bag.TalismanBag
import com.willfp.talismans.command.CommandTalismans
import com.willfp.talismans.display.TalismanDisplay
import com.willfp.talismans.libreforge.ConditionHasTalisman
import com.willfp.talismans.libreforge.SlotTypeTalisman
import com.willfp.talismans.talismans.TalismanTag
import com.willfp.talismans.talismans.Talismans
import com.willfp.talismans.talismans.util.BlockPlaceListener
import com.willfp.talismans.talismans.util.DiscoverRecipeListener
import com.willfp.talismans.talismans.util.TalismanChecks
import org.bukkit.entity.Player
import org.bukkit.event.Listener

internal lateinit var plugin: TalismansPlugin
    private set

class TalismansPlugin : LibreforgePlugin() {
    init {
        plugin = this

        TalismanChecks.registerItemStackProvider {
            TalismanBag.getTalismans(it)
        }

        TalismanChecks.registerItemStackProvider {
            if (configYml.getBool("read-inventory") && !configYml.getBool("offhand-only"))
                it.inventory.contents.filterNotNull()
            else emptyList()
        }

        TalismanChecks.registerItemStackProvider {
            if (configYml.getBool("read-enderchest") && !configYml.getBool("offhand-only"))
                @Suppress("UNNECESSARY_SAFE_CALL", "USELESS_ELVIS")
                it.enderChest?.contents?.filterNotNull() ?: emptyList()
            else emptyList()
        }
    }

    override fun handleLoad() {
        Items.registerTag(TalismanTag)
        Conditions.register(ConditionHasTalisman)
    }

    override fun handleEnable() {
        SlotTypes.register(SlotTypeTalisman)

        registerSpecificHolderProvider<Player> {
            TalismanChecks.getTalismansOnPlayer(it)
        }

        registerSpecificRefreshFunction<Player> {
            TalismanChecks.clearCache(it)
        }
    }

    override fun handleReload() {
        TalismanBag.update()
        TalismanChecks.reload()
    }

    override fun loadConfigCategories(): List<ConfigCategory> {
        return listOf(
            Talismans
        )
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandTalismans
        )
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            BlockPlaceListener,
            DiscoverRecipeListener
        )
    }

    override fun loadDisplayModules(): List<DisplayModule> {
        return listOf(
            TalismanDisplay
        )
    }

    override fun getCustomCharts() = listOf(
        EcoMetricsChart.SingleLine("total_talismans") { Talismans.values().size },
        EcoMetricsChart.SimplePie("discover_recipes") {
            if (configYml.getBool("crafting.discover")) "enabled" else "disabled"
        },
        EcoMetricsChart.SimplePie("read_inventory") {
            if (configYml.getBool("read-inventory")) "enabled" else "disabled"
        },
        EcoMetricsChart.SimplePie("read_enderchest") {
            if (configYml.getBool("read-enderchest")) "enabled" else "disabled"
        },
        EcoMetricsChart.SimplePie("read_shulkerboxes") {
            if (configYml.getBool("read-shulkerboxes")) "enabled" else "disabled"
        },
        EcoMetricsChart.SimplePie("top_level_only") {
            if (configYml.getBool("top-level-only")) "enabled" else "disabled"
        },
        EcoMetricsChart.SimplePie("offhand_only") {
            if (configYml.getBool("offhand-only")) "enabled" else "disabled"
        }
    )
}
